exp/zapslog/handler.go

```diff
@@ -157,29 +162,45 @@ func (h *Handler) Handle(ctx context.Context, record slog.Record) error {
 		ce.Stack = stacktrace.Take(3 + h.callerSkip)
 	}
 
-	fields := make([]zapcore.Field, 0, record.NumAttrs())
+	fields := make([]zapcore.Field, 0, record.NumAttrs()+1)
 	record.Attrs(func(attr slog.Attr) bool {
-		fields = append(fields, convertAttrToField(attr))
+		f := convertAttrToField(attr)
+		if h.holdGroup != "" && f != zap.Skip() {
+			fields = append(fields, zap.Namespace(h.holdGroup))
+			h.holdGroup = ""
+		}
+		fields = append(fields, f)
 		return true
 	})
+
 	ce.Write(fields...)
 	return nil
 }
 
 // WithAttrs returns a new Handler whose attributes consist of
 // both the receiver's attributes and the arguments.
 func (h *Handler) WithAttrs(attrs []slog.Attr) slog.Handler {
-	fields := make([]zapcore.Field, 0, len(attrs))
+	fields := make([]zapcore.Field, 0, len(attrs)+1)
 	for _, attr := range attrs {
-		fields = append(fields, convertAttrToField(attr))
+		f := convertAttrToField(attr)
+		if h.holdGroup != "" && f != zap.Skip() {
+			fields = append(fields, zap.Namespace(h.holdGroup))
+			h.holdGroup = ""
```

Comment:
Same here. Whether or not the namespace was added should be considered a concern local to this WithAttrs call.

After that, in the child handler constructed and returned below—that's where we unset the group because it's already been "used" here.

Reply:
fixed.